docs/docs/async-dialogs.md

```diff
@@ -4,3 +4,628 @@ title: Async Dialogs
+the `show()` and `hide()` methods. Its usage is more flexible and verbose:
+
+```javascript
+var output = document.getElementById('prompt');
+var prompt = new PromptDialog();
+
+prompt.show('What is your name?')
+  .waitForUser()
+  .then(function(name) {
+    output.innerHTML = '' + name;
+  })
+  .catch(function() {
+    output.innerHTML = '¯\\_(ツ)_/¯';
+  })
+  .finally(function() {
+    prompt.hide();
+  });
+```
+
+[Run example on JSBin](http://jsbin.com/wupixi/edit?js,output)
+
+This abstraction can be expanded on in other ways. For example a notification
+dialog:
+
+```javascript
+function NotifyDialog() {
+  Dialog.call(this);
+  var _this      = this;
+  this.el        = document.getElementById('notify-dialog');
+  this.messageEl = this.el.querySelector('.message');
+  this.okButton  = this.el.querySelector('button.ok');
+  this.okButton.addEventListener('click', function() {
+    _this._okCallback();
+  });
+}
+NotifyDialog.prototype = Object.create(Dialog.prototype);
+NotifyDialog.prototype.show = function(message) {
+  this.messageEl.innerHTML = '' + message;
+  this.el.className = '';
+  return this;
+};
+NotifyDialog.prototype.show = function() {
+  this.el.className = 'hidden';
+  return this;
+};
+```
+
+#### Exercises for the student
+
+1.  Write a function that takes a `Dialog` instance and a default value. Have it
+    return a promise that resolves to the default value if the user clicks
+    cancel.
+2.  With the use of abstract classes can the similarities between `PromptDialog`
+    and `NotifyDialog` be abstracted? Make a sub class of `Dialog` that
+    abstracts the common DOM code (`DOMDialog`). Then refactor the
+    `PromptDialog` and `NotifyDialog` to inherate from `DOMDialog` but
+    references the correct DOM selectors.
+
+## Cancellation
+
+Something missing from the above example is proper error handling. When it comes
+to promises it is a best practise to always *reject a promise with an Error* and
+not with plain data such as an object, string, number, or null/undefined. The
+reasoning for this is promises are best used as a way to regain some of the
+syntax you have with the standard `try {} catch() {}` blocks with asynchronous
+code.
+
+An advantage of using `Error`s is the ability to test why a promise was rejected
+and make decisions on that. This ability is also baked into how Bluebird works.
+You can pass in a predicate to the `catch()` block allowing you to have more
+than one block based on what `Error` it was rejected with. For example:
+
+```javascript
+doSomething().then(function(value) {
+  // Do something with value or fail with an error.
+  throw new Error('testing errors');
+})
+.catch(ArgumentError, function(e) {
+  console.log('You buggered up something with the arguments.', e);
+})
+.catch(SyntaxError, function(e) {
+  console.log('Check your syntax!', e);
+})
+.catch(function(e) {
+  // e is an Error object.
+  console.log('Well something genaric happened.', e);
```

Comment:
It might be a good idea to `throw e;` here as well to let unhandled rejection tracking show you the big stack trace.